server/src/conversation.ts

```diff
@@ -0,0 +1,290 @@
+import LruCache from "lru-cache";
+
+import pg from "./db/pg-query";
+import { MPromise } from "./utils/metered";
+
+import User from "./user";
+
+function createXidRecord(
+  ownerUid: any,
+  uid: any,
+  xid: any,
+  x_profile_image_url: any,
+  x_name: any,
+  x_email: any
+) {
+  return pg.queryP(
+    "insert into xids (owner, uid, xid, x_profile_image_url, x_name, x_email) values ($1, $2, $3, $4, $5, $6) " +
+      "on conflict (owner, xid) do nothing;",
+    [
+      ownerUid,
+      uid,
+      xid,
+      x_profile_image_url || null,
+      x_name || null,
+      x_email || null,
+    ]
+  );
+}
+
+function createXidRecordByZid(
+  zid: any,
+  uid: any,
+  xid: any,
+  x_profile_image_url: any,
+  x_name: any,
+  x_email: any
+) {
+  return getConversationInfo(zid).then(
+    (conv: { use_xid_whitelist: any; owner: any }) => {
+      const shouldCreateXidRecord = conv.use_xid_whitelist
+        ? isXidWhitelisted(conv.owner, xid)
+        : Promise.resolve(true);
+      return shouldCreateXidRecord.then((should: any) => {
+        if (!should) {
+          throw new Error("polis_err_xid_not_whitelisted_2");
+        }
+        return pg.queryP(
+          "insert into xids (owner, uid, xid, x_profile_image_url, x_name, x_email) values ((select org_id from conversations where zid = ($1)), $2, $3, $4, $5, $6) " +
+            "on conflict (owner, xid) do nothing;",
+          [
+            zid,
+            uid,
+            xid,
+            x_profile_image_url || null,
+            x_name || null,
+            x_email || null,
+          ]
+        );
+      });
+    }
+  );
+}
+
+function getXidRecord(xid: any, zid: any) {
+  return pg.queryP(
+    "select * from xids where xid = ($1) and owner = (select org_id from conversations where zid = ($2));",
+    [xid, zid]
+  );
+}
+
+function getXidRecordByXidOwnerId(
+  xid: any,
+  owner: any,
+  zid_optional: any,
+  x_profile_image_url: any,
+  x_name: any,
+  x_email: any,
+  createIfMissing: any
+) {
+  return (
+    pg
+      .queryP("select * from xids where xid = ($1) and owner = ($2);", [
+        xid,
+        owner,
+      ])
+      //     (local function)(rows: string | any[]): any
+      // Argument of type '(rows: string | any[]) => any' is not assignable to parameter of type '(value: unknown) => any'.
+      //   Types of parameters 'rows' and 'value' are incompatible.
+      //     Type 'unknown' is not assignable to type 'string | any[]'.
+      //     Type 'unknown' is not assignable to type 'any[]'.ts(2345)
+      // @ts-ignore
+      .then(function (rows: string | any[]) {
```

Comment:
When I couldn't quickly figure out a good way to solve a TypeScript error, I pasted the error as a comment and then added a magic `// @ts-ignore` comment so that a domain expert can come back later, in another PR, and fix the error.

Often in these cases I suspect that the fixes could involve some code refactoring. When that was the case, I opted to ignore the error for now, and then suggest that a future contributor attempt the refactor later, in isolation, in order to avoid doing to many things at once and breaking code that I am not too familiar with yet.

Reply:
This particular class of TypeScript errors that have to do with Promises are something worth looking into.  I need to read up more on how to specify the return type for a Promise.

Maybe useful: 

https://www.jpwilliams.dev/how-to-unpack-the-return-type-of-a-promise-in-typescript

https://itnext.io/typescript-with-promises-and-async-await-63623b8e5e2a

https://stackoverflow.com/questions/50826605/typescript-type-function-returning-a-promise